File: app/backend/java/auth-service/src/main/java/com/srbms/authservice/service/AuthService.java
package com.srbms.authservice.service;

import com.srbms.authservice.dto.JwtResponse;
import com.srbms.authservice.dto.LoginRequest;
import com.srbms.authservice.dto.RegisterRequest;
import com.srbms.authservice.entity.Driver;
import com.srbms.authservice.entity.Role;
import com.srbms.authservice.entity.User;
import com.srbms.authservice.repository.DriverRepository;
import com.srbms.authservice.repository.RoleRepository;
import com.srbms.authservice.repository.UserRepository;
import com.srbms.authservice.security.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.Optional;

@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RoleRepository roleRepository;

    @Autowired
    private DriverRepository driverRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtil jwtUtil;

    public JwtResponse login(LoginRequest loginRequest) {
        String identifier = loginRequest.getEmailOrUsername();
        if (identifier == null || identifier.isBlank()) {
            throw new RuntimeException("Mobile number or username is required.");
        }

        // Try searching by phone, then username, then email
        Optional<User> userOpt = userRepository.findByPhone(identifier);
        if (userOpt.isEmpty()) {
            userOpt = userRepository.findByUsername(identifier);
        }
        if (userOpt.isEmpty() && identifier.contains("@")) {
            userOpt = userRepository.findByEmail(identifier);
        }

        if (userOpt.isEmpty()) {
            throw new RuntimeException("Invalid mobile number/username or password");
        }

        User user = userOpt.get();

        // Support both hashed passwords and legacy plain text from dump
        boolean passwordMatches = passwordEncoder.matches(loginRequest.getPassword(), user.getPassword())
                || loginRequest.getPassword().equals(user.getPassword());

        if (!passwordMatches) {
            throw new RuntimeException("Invalid mobile number/username or password");
        }

        String userStatus = user.getStatus() != null ? user.getStatus().toLowerCase() : "active";
        if ("inactive".equals(userStatus) || "deactivated".equals(userStatus) || "disabled".equals(userStatus) || "suspended".equals(userStatus)) {
            throw new RuntimeException("Your account is currently Inactive. Please contact customer support to reactivate your account.");
        }

        String roleName = user.getRole() != null ? user.getRole().getRoleValue() : "rider";
        String token = jwtUtil.generateToken(user.getUserId(), user.getUsername(), user.getEmail(), roleName);

        return new JwtResponse(token, user.getUserId(), user.getUsername(), user.getEmail(), user.getPhone(), roleName);
    }

    @org.springframework.transaction.annotation.Transactional
    public User register(RegisterRequest registerRequest) {
        String rawPhone = registerRequest.getPhone();
        if (rawPhone == null || rawPhone.isBlank()) {
            throw new RuntimeException("Mobile number is required for registration!");
        }

        String phone = rawPhone.replaceAll("^\\+91", "").replaceAll("[^0-9]", "");
        if (userRepository.existsByPhone(phone) || userRepository.existsByPhone("+91" + phone)) {
            throw new RuntimeException("Mobile number +91" + phone + " is already registered! Please use a different number or login.");
        }

        // Username defaults to phone number if not explicitly set
        String username = registerRequest.getUsername();
        if (username == null || username.isBlank()) {
            username = phone;
        } else {
            username = username.trim();
        }

        if (userRepository.existsByUsername(username)) {
            throw new RuntimeException("Username '" + username + "' is already taken! Please choose a different username.");
        }

        // Handle optional email
        String email = registerRequest.getEmail();
        if (email != null && email.isBlank()) {
            email = null;
        } else if (email != null) {
            email = email.trim();
        }
        if (email != null && userRepository.existsByEmail(email)) {
            throw new RuntimeException("Email '" + email + "' is already registered! Please use a different email address.");
        }

        String requestedRole = registerRequest.getRole() != null ? registerRequest.getRole().toLowerCase() : "rider";
        Role role = roleRepository.findByRoleValue(requestedRole)
                .orElseGet(() -> roleRepository.save(new Role(requestedRole)));

        User user = new User();
        user.setUsername(username);
        user.setEmail(email);
        user.setPassword(passwordEncoder.encode(registerRequest.getPassword()));
        user.setPhone(phone);
        user.setDob(registerRequest.getDob());
        user.setGender(registerRequest.getGender());
        user.setRole(role);
        if ("driver".equalsIgnoreCase(requestedRole)) {
            user.setStatus("Pending Verification");
        } else {
            user.setStatus("active");
        }
        user.setProfileImage(registerRequest.getProfileImage() != null ? registerRequest.getProfileImage() : "default.jpg");

        User savedUser = userRepository.save(user);

        if ("driver".equalsIgnoreCase(requestedRole)) {
            if (registerRequest.getLicenseNo() == null || registerRequest.getLicenseNo().isBlank()) {
                throw new RuntimeException("Driver license number is required for driver registration!");
            }
            Driver driver = new Driver(savedUser, registerRequest.getLicenseNo(), "unverified", registerRequest.getLicensePdfUrl());
            driverRepository.save(driver);
        }

        return savedUser;
    }

    public Map<String, Object> validateToken(String token) {
        if (token != null && token.startsWith("Bearer ")) {
            token = token.substring(7);
        }

        boolean isValid = jwtUtil.validateToken(token);
        if (!isValid) {
            return Map.of("valid", false);
        }

        return Map.of(
            "valid", true,
            "userId", jwtUtil.extractUserId(token),
            "email", jwtUtil.extractSubject(token),
            "role", jwtUtil.extractRole(token),
            "username", jwtUtil.extractUsername(token)
        );
    }
}

File: app/backend/.net/SmartRide.AuthService/Services/AuthService.cs
using System.Text.RegularExpressions;
using Microsoft.EntityFrameworkCore;
using SmartRide.AuthService.Data;
using SmartRide.AuthService.Models.DTOs;
using SmartRide.AuthService.Models.Entities;

namespace SmartRide.AuthService.Services;

public class AuthService : IAuthService
{
    private readonly AuthDbContext _context;
    private readonly IJwtService _jwtService;

    public AuthService(AuthDbContext context, IJwtService jwtService)
    {
        _context = context;
        _jwtService = jwtService;
    }

    public async Task<JwtResponse> LoginAsync(LoginRequest loginRequest)
    {
        string identifier = loginRequest.EmailOrUsername ?? string.Empty;
        if (string.IsNullOrWhiteSpace(identifier))
        {
            throw new InvalidOperationException("Mobile number or username is required.");
        }

        // Try searching by phone, then username, then email
        User? user = await _context.Users
            .Include(u => u.Role)
            .FirstOrDefaultAsync(u => u.Phone == identifier);

        if (user == null)
        {
            user = await _context.Users
                .Include(u => u.Role)
                .FirstOrDefaultAsync(u => u.Username == identifier);
        }

        if (user == null && identifier.Contains('@'))
        {
            user = await _context.Users
                .Include(u => u.Role)
                .FirstOrDefaultAsync(u => u.Email == identifier);
        }

        if (user == null)
        {
            throw new InvalidOperationException("Invalid mobile number/username or password");
        }

        bool passwordMatches = false;
        try
        {
            passwordMatches = BCrypt.Net.BCrypt.Verify(loginRequest.Password, user.Password);
        }
        catch
        {
            // Fallback for legacy plain text passwords in database
            passwordMatches = (loginRequest.Password == user.Password);
        }

        if (!passwordMatches && loginRequest.Password == user.Password)
        {
            passwordMatches = true;
        }

        if (!passwordMatches)
        {
            throw new InvalidOperationException("Invalid mobile number/username or password");
        }

        string roleName = user.Role != null ? user.Role.RoleValue : "rider";
        string token = _jwtService.GenerateToken(user.UserId, user.Username, user.Email, roleName);

        return new JwtResponse(token, user.UserId, user.Username, user.Email, user.Phone, roleName);
    }

    public async Task<User> RegisterAsync(RegisterRequest registerRequest)
    {
        string rawPhone = registerRequest.Phone ?? string.Empty;
        if (string.IsNullOrWhiteSpace(rawPhone))
        {
            throw new InvalidOperationException("Mobile number is required for registration!");
        }

        string phone = Regex.Replace(rawPhone, @"^\+91", "");
        phone = Regex.Replace(phone, @"[^0-9]", "");

        bool phoneExists = await _context.Users.AnyAsync(u => u.Phone == phone || u.Phone == "+91" + phone);
        if (phoneExists)
        {
            throw new InvalidOperationException($"Mobile number +91{phone} is already registered! Please use a different number or login.");
        }

        string username = registerRequest.Username ?? string.Empty;
        if (string.IsNullOrWhiteSpace(username))
        {
            username = phone;
        }
        else
        {
            username = username.Trim();
        }

        bool usernameExists = await _context.Users.AnyAsync(u => u.Username == username);
        if (usernameExists)
        {
            throw new InvalidOperationException($"Username '{username}' is already taken! Please choose a different username.");
        }

        string? email = registerRequest.Email;
        if (string.IsNullOrWhiteSpace(email))
        {
            email = null;
        }
        else
        {
            email = email.Trim();
        }

        if (email != null && await _context.Users.AnyAsync(u => u.Email == email))
        {
            throw new InvalidOperationException($"Email '{email}' is already registered! Please use a different email address.");
        }

        string requestedRole = registerRequest.Role != null ? registerRequest.Role.ToLower() : "rider";
        Role? role = await _context.Roles.FirstOrDefaultAsync(r => r.RoleValue.ToLower() == requestedRole);
        if (role == null)
        {
            role = new Role(requestedRole);
            _context.Roles.Add(role);
            await _context.SaveChangesAsync();
        }

        string hashedPassword = BCrypt.Net.BCrypt.HashPassword(registerRequest.Password ?? string.Empty);

        var user = new User
        {
            Username = username,
            Email = email,
            Password = hashedPassword,
            Phone = phone,
            Dob = registerRequest.Dob ?? new DateTime(1990, 1, 1),
            Gender = registerRequest.Gender,
            RoleId = role.RoleId,
            Role = role,
            Status = "active",
            ProfileImage = registerRequest.ProfileImage ?? "default.jpg"
        };

        _context.Users.Add(user);
        await _context.SaveChangesAsync();

        if ("driver".Equals(requestedRole, StringComparison.OrdinalIgnoreCase))
        {
            if (string.IsNullOrWhiteSpace(registerRequest.LicenseNo))
            {
                throw new InvalidOperationException("Driver license number is required for driver registration!");
            }

            var driver = new Driver(user, registerRequest.LicenseNo, "unverified", registerRequest.LicensePdfUrl);
            _context.Drivers.Add(driver);
            await _context.SaveChangesAsync();
        }

        return user;
    }

    public IDictionary<string, object> ValidateToken(string token)
    {
        if (!string.IsNullOrWhiteSpace(token) && token.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
        {
            token = token.Substring(7).Trim();
        }

        bool isValid = _jwtService.ValidateToken(token);
        if (!isValid)
        {
            return new Dictionary<string, object> { { "valid", false } };
        }

        return new Dictionary<string, object>
        {
            { "valid", true },
            { "userId", _jwtService.ExtractUserId(token) },
            { "email", _jwtService.ExtractSubject(token) },
            { "role", _jwtService.ExtractRole(token) },
            { "username", _jwtService.ExtractUsername(token) }
        };
    }
}
